app/com/linkedin/drelephant/tuning/engine/SparkHBTParamRecommender.java

```diff
@@ -0,0 +1,361 @@
+   * executor memory, core, driver memory and memory factor
+   *
+   * @return HashMap with property name as key and double value
+   */
+  public HashMap<String, Double> getHBTSuggestion() {
+    HashMap<String, Double> suggestedParameters = new HashMap<String, Double>();
+    try {
+      suggestExecutorMemoryCore();
+      suggestMemoryFactor();
+      suggestDriverMemory();
+      suggestedParameters.put(SparkConfigurationConstants.SPARK_EXECUTOR_MEMORY_KEY, (double) suggestedExecutorMemory
+          / FileUtils.ONE_MB);
+      suggestedParameters.put(SparkConfigurationConstants.SPARK_EXECUTOR_CORES_KEY, suggestedCore.doubleValue());
+      if (suggestedMemoryFactor < UnifiedMemoryHeuristic.SPARK_MEMORY_FRACTION_THRESHOLD()) {
+        suggestedMemoryFactor = UnifiedMemoryHeuristic.SPARK_MEMORY_FRACTION_THRESHOLD();
+      }
+      suggestedParameters.put(SparkConfigurationConstants.SPARK_MEMORY_FRACTION_KEY, suggestedMemoryFactor);
+      suggestedParameters.put(SparkConfigurationConstants.SPARK_DRIVER_MEMORY_KEY, (double) suggestedDriverMemory
+          / FileUtils.ONE_MB);
+      logger.info("Following are the suggestions for spark parameters for app id : " + appResult.flowExecId);
+      logger.info("suggestedExecutorMemory " + suggestedExecutorMemory);
+      logger.info("suggestedCore " + suggestedCore);
+      logger.info("suggestedMemoryFactor " + suggestedMemoryFactor);
+      logger.info("suggestedDriverMemory " + suggestedDriverMemory);
+    } catch (Exception e) {
+      logger.error("Error in generating parameters ", e);
+    }
+    return suggestedParameters;
+  }
+
+  private Long getMaxPeakJVMUsedMemoryPerCore() {
+    return maxPeakJVMUsedMemory / lastRunExecutorCore;
+  }
+
+  private Long getMaxPeakUnifiedMemoryPerCore() {
+    return maxPeakUnifiedMemory / lastRunExecutorCore;
+  }
+
+  /**
+   * Suggest executor memory and core We calculate per core peak JVM used memory. Then we start from
+   * max executor core and compute memory requirement for max executor core If this is below
+   * threshold of max executor core then good, otherwise reduce the core and check memory
+   * requirement.
+   */
+  private void suggestExecutorMemoryCore() {
+    boolean validSuggestion = false;
+    long maxPeakJVMUsedMemoryPerCore = getMaxPeakJVMUsedMemoryPerCore();
+    for (int core = MAX_EXECUTOR_CORE; core > 0; core--) {
+      long currSuggestedMemory = suggestExecutorMemory(maxPeakJVMUsedMemoryPerCore, core);
+      if (currSuggestedMemory < MAX_EXECUTOR_MEMORY || core == 1) {
+        suggestedExecutorMemory = currSuggestedMemory;
+        suggestedCore = core;
+        validSuggestion = true;
+        break;
+      }
+    }
+    if (!validSuggestion) {
+      suggestedExecutorMemory = lastRunExecutorMemory;
+      suggestedCore = lastRunExecutorCore;
+    }
+    if (getMemoryIncreaseForGC() != 0) {
+      suggestedExecutorMemory = suggestedExecutorMemory * (100 + getMemoryIncreaseForGC()) / 100;
```

Comment:
After increasing suggested memory , we should check if it increase beyond MAX_EXECUTOR_MEMORY